lib/r10k/tarball.rb

```diff
@@ -0,0 +1,250 @@
+      end
+      purge ? target_dir_entries.empty? : true
+    end
+
+    # Download the tarball from @source to @cache_path
+    def download
+      Dir.mktmpdir do
+        tempfile = Tempfile.new(tarball_basename)
+        tempfile.binmode
+        src_uri = URI.parse(source)
+
+        temp_digest = case src_uri.scheme
+                      when 'file', nil
+                        copy_to_file(src_uri.path, tempfile)
+                      when %r{https?}
+                        download_to_file(src_uri, tempfile)
+                      else
+                        raise "Unexpected source scheme #{src_uri.scheme}"
+                      end
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
+          raise 'Downloaded file does not match digest'
+        end
+
+        # Move the download to cache_path
+        FileUtils::mkdir_p(cache_root)
+        begin
+          FileUtils.mv(tempfile.path, cache_path)
+        rescue Errno::EACCES
+          FileUtils.cp(tempfile.path, cache_path)
+        end
+
+        tempfile.close
+        tempfile.unlink
+      end
+    end
+
+    # Is the cached tarball valid? Checks the cached tarball's digest against
+    # the expected digest.
+    #
+    # @return [Boolean]
+    def cache_valid?
+      unless sha256digest.nil?
+        File.exist?(cache_path) && sha256digest == file_digest(cache_path)
+      else
+        File.exist?(cache_path)
+      end
+    end
+
+    # List all of the files contained in the tarball and their paths. This is
+    # useful for implementing R10K::Purgable
+    #
+    # @return [Array] A list of file paths contained in the archive
+    def paths
+      names = Array.new
+      each_tarball_entry { |entry| names << entry.full_name.chomp('/') }
+      names
+    end
+
+    private
+
+    CHUNK_SIZE = 64 * 1024 # 64 kb
+
+    def each_tarball_entry(&block)
+      File.open(cache_path, 'rb') do |file|
+        Zlib::GzipReader.wrap(file) do |reader|
+          Archive::Tar::Minitar::Input.each_entry(reader) do |entry|
+            yield entry
+          end
+        end
+      end
+    end
+
+    # Return the sha256 digest of the file at the given path
+    #
+    # @param path [String] The path to the file
+    # @return [String] The file's sha256 hex digest
+    def file_digest(path)
+      File.open(path) do |file|
+        reader_digest(file)
+      end
+    end
+
+    # Return the sha256 digest of the readable data
+    #
+    # @param reader [String] An object that responds to #read
+    # @return [String] The read data's sha256 hex digest
+    def reader_digest(reader)
+      digest = Digest::SHA256.new
+      while chunk = reader.read(CHUNK_SIZE)
+        digest.update(chunk)
+      end
+
+      digest.hexdigest
+    end
+
+    def sanitized_dirname(string)
+      string.gsub(/[^@\w\.-]/, '-')
+    end
+
+    def http_get(uri, method: Net::HTTP::Get, redirect_limit: 10, &block)
```

Comment:
If you want this to be a generic http request method, I would change the method name to reflect that. Otherwise, if you want to keep it specific to GET, I would remove `method` as a param.

This also seems like something that should be shared, rather than specific to this class.